traffic_ops/traffic_ops_golang/server/servers_server_capability.go

```diff
@@ -453,77 +454,166 @@ func AssignMultipleServerCapabilities(w http.ResponseWriter, r *http.Request) {
-	}
-	if !exists {
-		userErr := fmt.Errorf("server %d does not exist", msc.ServerID)
-		api.HandleErr(w, r, tx, http.StatusNotFound, userErr, nil)
-		return
+	if len(mssc.ServerIDs) == 1 {
+		errCode, userErr, sysErr = checkExistingServer(tx, mssc.ServerIDs[0], inf.User.UserName)
+		if userErr != nil || sysErr != nil {
+			api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+			return
+		}
 	}
 
-	// Ensure type is correct
-	correctType := true
-	if err := tx.QueryRow(scCheckServerTypeQuery(), msc.ServerID).Scan(&correctType); err != nil {
+	//Check if the server type is MID and/or EDGE
+	var servArray []int64
+	queryType := `SELECT array_agg(s.id) 
+		FROM server s
+		JOIN type t ON s.type = t.id
+		WHERE s.id = any ($1)
+		AND t.use_in_table = 'server'
+		AND (t.name LIKE 'MID%' OR t.name LIKE 'EDGE%')`
+	if err := tx.QueryRow(queryType, pq.Array(mssc.ServerIDs)).Scan(pq.Array(&servArray)); err != nil {
 		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, fmt.Errorf("checking server type: %w", err))
 		return
 	}
-	if !correctType {
-		userErr := fmt.Errorf("server %d has an incorrect server type. Server capabilities can only be assigned to EDGE or MID servers", msc.ServerID)
-		api.HandleErr(w, r, tx, http.StatusBadRequest, userErr, nil)
-		return
+	cmp := make(map[int64]bool)
+	for _, item := range servArray {
+		cmp[item] = true
+	}
+	for _, sid := range mssc.ServerIDs {
+		if _, ok := cmp[sid]; !ok {
+			userErr := fmt.Errorf("server id: %d has an incorrect server type. Server capability can only be assigned to EDGE or MID servers", sid)
+			api.HandleErr(w, r, tx, http.StatusBadRequest, userErr, nil)
+			return
+		}
 	}
 
-	cdnName, err := dbhelpers.GetCDNNameFromServerID(tx, int64(msc.ServerID))
+	// Insert rows in DB
+	sid := make([]int64, len(mssc.ServerCapabilities))
+	scs := make([]string, len(mssc.ServerIDs))
+	if len(mssc.ServerIDs) == 1 {
+		if len(mssc.ServerCapabilities) >= 1 {
+			for i := range mssc.ServerCapabilities {
+				sid[i] = mssc.ServerIDs[0]
+			}
+			scs = mssc.ServerCapabilities
+		}
+	} else if len(mssc.ServerCapabilities) == 1 {
+		if len(mssc.ServerIDs) >= 1 {
+			for i := range mssc.ServerIDs {
+				scs[i] = mssc.ServerCapabilities[0]
+			}
+			sid = mssc.ServerIDs
+		}
+	} else {
+		scs = mssc.ServerCapabilities
+		sid = mssc.ServerIDs
+	}
+
+	msscQuery := `INSERT INTO server_server_capability
+			select "server_capability", "server"
+			FROM UNNEST($1::text[], $2::int[]) AS tmp("server_capability", "server")`
+	_, err := tx.Query(msscQuery, pq.Array(scs), pq.Array(sid))
 	if err != nil {
-		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, err)
+		useErr, sysErr, statusCode := api.ParseDBError(err)
+		api.HandleErr(w, r, tx, statusCode, useErr, sysErr)
 		return
 	}
 
-	userErr, sysErr, errCode = dbhelpers.CheckIfCurrentUserCanModifyCDN(tx, string(cdnName), inf.User.UserName)
+	var alerts tc.Alerts
+	if len(mssc.ServerCapabilities) == 1 && len(mssc.ServerIDs) == 1 {
```

Comment:
I was thinking some more about this last night after we spoke. Do you think it would make sense to have another query parameter added to the request that says whether the assignment is from a server to capability or the other way around. Then based on that query param, you could form your database queries. That way, we could get rid of all the extra `if..else` conditions.

Reply:
Not really, since you still have to figure out whether it is mass-assignment or just one association or deletion.

Reply:
But that's what I mean by an extra parameter. So, say that we have our logic so that if it comes from the servers page, we have that value set to true, else we have it set to false. That should do it, no? And by default, we have it set to true maybe?